numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -1898,6 +1916,16 @@ PyArray_AddCastingImplmentation(PyBoundArrayMethodObject *meth)
                     meth->method->name);
             return -1;
         }
+        if (meth->dtypes[0]->within_dtype_castingimpl != NULL) {
+            PyErr_Format(PyExc_RuntimeError,
+                    "A cast was already added for %S -> %S. (method: %s)",
+                    meth->dtypes[0], meth->dtypes[1], meth->method->name);
+            return -1;
```

Comment:
If this could be hit by a user-defined type (as I think it could), it would be good to test for

Reply:
It can't be hit by a current user defined dtype. I could currently only hit it by adding a cast twice to a NumPy builtin DType.

Reply:
Please add a comment then that it is an internal error, and will never be hit in practice